source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown

```diff
@@ -0,0 +1,131 @@
+
+By creating this a new `/riders` endpoint, this new resource can have a brand new representation. As always, the same database fields can be used internally, the same internal alerting tools are used for letting folks know about matches (v1 app) or new pending riders (v2 app). The API can create and update "matches" through this new "riders" interface. Clients can then use either one, and the code just figures itself out in the background. Over time the refactoring can be done to move the internal logic more towards riders, and your [integration tests / contract tests](https://blog.apisyouwonthate.com/tricking-colleagues-into-writing-documentation-via-contract-testing-206308b47e05) will confirm that things aren't changing on the outside.
+
+> We have all these old endpoints hanging around, can we get rid of these slightly more intelligently than just sending some emails?
+
+Yes! API Endpoints can be marked with a `Sunset` header to signal deprecation (and eventual removal) of an endpoint.
+
+The [Sunset header][sunset-draft] is an in-development HTTP response header that is aiming to standardize how URLs are marked for deprecation. tl;dr it looks a bit like this:
+
+```
+Sunset: Sat, 31 Dec 2018 23:59:59 GMT
+```
+
+The date is a [HTTP date], and can be combined with a `Link: <http://foo.com/something> rel="sunset"` which can be anything that might help a developer know what is going on. Maybe link to your API documentation for the new resource, the OpenAPI/JSON Schema definitions, or even a blog post explaining the change.
+
+Ruby on Rails has [rails-sunset], and hopefully other frameworks will start adding this functionality. Open-source API Gateway system [Tyk](https://tyk.io/) is [adding support to an upcoming version](https://github.com/TykTechnologies/tyk/issues/1626).
+
+Clients then add a middleware to their HTTP calls, checking for Sunset headers. We do this with [faraday-sunset](https://github.com/wework/faraday-sunset) (Ruby), [Hunter Skrasek](https://twitter.com/HSkrasek/) made [guzzle-sunset](https://github.com/hskrasek/guzzle-sunset) (PHP), and anyone can write a thing that looks for a header and logs it to whatever logging thing they're using.
+
+[HTTP date]: https://tools.ietf.org/html/rfc7231#section-7.1.1.1
+[sunset-draft]: https://tools.ietf.org/html/draft-wilde-sunset-header-03
+[rails-sunset]: https://github.com/wework/rails-sunset/
+
+> Changing validation rules, something seemingly backwards compatible, like making the length of a field a little longer.
+
+Changing validation rules seems like it might be a backwards compatible change, but often it is not. For example, making a string field accept a longer value can lead to some interesting problems. As clients often bake validation rules into client applications based on whatever the documentation says, if that documentation changes they aren't going to notice, and that validation will mismatch on different clients.
```

Comment:
This isn't very convincing as backwards incompatible; this particular example won't "break" any clients, just they won't benefit from the additional available length. Maybe in addition, provide an example of wanting to introduce more restrictive validation which is definitely breaking.

Reply:
So, maybe this example could be change to a scenario where the maximum length is decreased?

Reply:
@josephgardner  It can break clients.  Client libraries are often built to allow you to GET an object, make some modifications and then PUT it back.  If you are using a down-rev client and retrieve an object that has a description that is longer than the old constraint, the GET/change/PUT flow will fail.

Reply:
There is an even more insidious issue with optional fields which all assume are non-breaking.  However, what happens when an up-rev client sets an option value and then a down-rev client that knows nothing about the optional value, does a GET/change/PUT flow.  The update will not contain the optional value and the PUT semantics cause the optional value to be deleted.  So far the only solution I have found for this is to version the message.

Reply:
… or do user-agent sniffing  (ick)

Reply:
Ahh, well if you read the attached link it's absolutely a breaking change and will royally fuck things up.

Reply:
Updated this in 06712a47d, thank you !